torchrl/objectives/dqn.py

```diff
@@ -560,7 +560,7 @@ def forward(self, input_tensordict: TensorDictBase) -> TensorDict:
             support = support.to("cpu")
             pns_a = pns_a.to("cpu")
 
-            Tz = reward + (1 - terminated.to(reward.dtype)) * discount * support
+            Tz = reward + (1 - terminated.to(reward.dtype)) * discount.unsqueeze(-1) * support.repeat(batch_size, 1)
```

Comment:
Looks good
I think we're making a couple of assumptions about discount and support here.
I would feel safer if we had a block before this where we check explicitly the shapes:
```python
if not check(discount.shape):
    raise RuntimeError(f"Expected `discount` to have either shape X or Y, got {discount.shape}")
if not other_check(support.shape):
    raise RuntimeError(f"Expected `support` to have either shape X or Y, got {support.shape}")
```
(`check` and `other_check` are just placeholders)
If possible I would also make an instance of this loss class with the support and discount shape that you are using to test that it works ok (see `test_cost.py:TestDQN:test_distributional_dqn`)

Reply:
We should also make sure that discount is a tensor:
```python
discount = torch.as_tensor(discount)
```
should do it (it's a no-op if it's already a tensor)